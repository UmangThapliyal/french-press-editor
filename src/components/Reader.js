//
// `<Reader />` component,
// a read-only version of the Slate Editor
//
// tools
import React from "react"
import { Editor } from "slate-react"
import { Value } from "slate"
import { schema } from "../schema"
import { renderNode, renderMark } from "../render"
//
// component export
export const Reader = props => {
  <Editor
    // this is the Slate editor component that renders Slate JSON value object
    //
    // props defineable by user
    options={{
      ...props.options,
      imagePlaceholder:
        (props.options && props.options.imagePlaceholder) ||
        "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7",
      domain: (props.options && props.options.domain) || ""
    }}
    components={props.components}
    value={Value.fromJSON(props.value)}
    //
    // props inferred from Editor architecture
    readOnly={true}
    schema={schema}
    renderNode={renderNode}
    renderMark={renderMark}
  />
}
