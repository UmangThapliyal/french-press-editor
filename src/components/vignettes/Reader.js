import React from "react"

import {
  RULES_SERIALIZE,
  rulesSerializeWithProps,
} from "../../constants/rules-serialize"

const rules = props => [
  ...RULES_SERIALIZE,
  ...rulesSerializeWithProps(props),
  {
    serialize(node, children) {
      if (node.object === "string") {
        return children.split("\n").reduce((array, text, i) => {
          if (i !== 0) array.push(<br key={i} />)
          array.push(text)
          return array
        }, [])
      }
    },
  },
]

export const addRootSerialNumbers = nodes => {
  nodes.forEach((node, iterator) => {
    node.serial = iterator
  })
  return nodes
}
let nodes = []
export const Reader = props => {
  const {document} = props.value
  nodes = addRootSerialNumbers(document.nodes)
  const elements = nodes.map(serializeNode).filter(Boolean)
  return elements
}
export const serializeNode = node => {
  if (node.object === "text") {
    const {leaves} = node
    return leaves.map(serializeLeaf)
  }
  const children = node.nodes.map(serializeNode)
  for (const rule of rules) {
    if (!rule.serialize) continue
    const ret = rule.serialize(node, children)
    if (ret === null) return
    if (ret) return addKey(ret)
  }
  throw new Error(`No serializer defined for node of type "${node.type}".`)
}
export const serializeLeaf = leaf => {
  const string = {object: "string", text: leaf.text}
  const text = serializeString(string)
  if (!leaf.marks) return leaf.text

  return leaf.marks.reduce((children, mark) => {
    for (const rule of rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(mark, children)
      if (ret === null) return null
      if (ret) return addKey(ret)
    }
    throw new Error(`No serializer defined for mark of type "${mark.type}".`)
  }, text)
}
export const serializeString = string => {
  for (const rule of rules) {
    if (!rule.serialize) continue
    const ret = rule.serialize(string, string.text)
    if (ret) return ret
  }
}
let key = 0
export const addKey = element => {
  const thisKey = key++
  return React.cloneElement(element, {
    key: thisKey,
    node: element.props.node
      ? {
          ...element.props.node,
          get: object => {
            return object === "key" && thisKey
          },
        }
      : null,
    parent: {
      getNextBlock: () => {
        const node = nodes.filter(object => {
          return object.serial === element.props.node.serial + 1
        })[0]
        const nodeFunction = {
          get: object =>
            object === "data" && {
              get: object => (node.data ? node.data[object] : undefined),
            },
        }
        return nodeFunction
      },
    },
  })
}

export default Reader
