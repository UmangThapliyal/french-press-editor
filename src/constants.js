export const DEFAULT_EDITOR_STATE = {
  document: {
    nodes: [
      {
        object: "block",
        type: "paragraph",
        nodes: [
          {
            object: "text",
            leaves: [
              {
                text: ""
              }
            ]
          }
        ]
      }
    ]
  }
}

export const PLACEHOLDER_TEXT = "Write your storyâ€¦"

// accepted upload file types
export const PICTURE_ACCEPTED_UPLOAD_MIME = ["image/png", "image/jpeg"]
// write the above in human language for warnign boxes
export const PICTURE_ACCEPTED_UPLOAD_MIME_HUMAN = "PNG or JPEG"
