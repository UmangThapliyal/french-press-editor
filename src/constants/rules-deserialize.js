/**
 * A set of rules that defines how to transpile HTML itno slate document nodes and marks (and back).
 * @constant rules
 */
import isUrl from "is-url"

import {BLOCK_TAGS, MARK_TAGS} from "./defaults"
import {squish} from "../utils/squish-to-plaintext"

export const hasImages = nodes => {
  let hasImages = false
  nodes.forEach(el => {
    if (el.tagName) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      hasImages = block === "image"
    }
  })
  return hasImages
}

export const imageDeserializer = el => {
  const imageSrc = el.getAttribute("src") || el.getAttribute("srcset")
  if (!isUrl(imageSrc)) return
  return {
    object: "block",
    type: "image",
    isVoid: true,
    data: {src: imageSrc},
  }
}

export const RULES_DESERIALIZE = [
  {
    deserialize(el, next) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      if (!block) return
      switch (block) {
        case "paragraph": {
          if (!hasImages(el.childNodes))
            return {
              object: "block",
              type: "paragraph",
              nodes: next(el.childNodes),
            }
          else return next(el.childNodes)
        }
        case "quote": {
          return {
            object: "block",
            type: "quote",
            nodes: next(squish(el).childNodes),
          }
        }
        case "heading": {
          return {
            object: "block",
            type: "heading",
            nodes: next(squish(el).childNodes),
          }
        }
        case "image":
          return imageDeserializer(el)
        case "link": {
          return {
            object: "inline",
            type: "link",
            data: {
              href: el.getAttribute("href"),
            },
            nodes: next(squish(el).childNodes),
          }
        }
        default:
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
      }
    },
  },
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName]
      if (!mark) return
      return {
        object: "mark",
        type: mark,
        nodes: next(el.childNodes),
      }
    },
  },
]
