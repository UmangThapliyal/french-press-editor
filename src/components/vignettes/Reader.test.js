import "jest-styled-components"

import React from "react"

import {shallow} from "enzyme"

import Reader, {addRootSerialNumbers, rules, nodeFactory} from "./Reader"
import {EXAMPLE_VALUE} from "../../../examples/constants"

test("rules() with no props creates three serialize objects in an array", () => {
  expect(rules({})).toHaveLength(3)
})

test("addRootSerialNumbers adds serial numbers to (shallow) node objects", () => {
  const nodes = [
    {
      object: "block",
      type: "paragraph",
      nodes: [{object: "text"}],
    },
    {
      object: "block",
      type: "heading",
    },
  ]
  expect(addRootSerialNumbers(nodes)).toEqual([
    {
      nodes: [{object: "text"}],
      object: "block",
      serial: 0,
      type: "paragraph",
    },
    {
      object: "block",
      serial: 1,
      type: "heading",
    },
  ])
})

test("nodeFactory maps Value nodes onto serializeNode function instances, produces six blocks", () => {
  const serializeNode = jest.fn()
  expect(nodeFactory({value: EXAMPLE_VALUE})).toHaveLength(6)
})

test("Render Reader without crashing, matches snapshot", () => {
  const element = shallow(<Reader />)
  expect(element).toMatchSnapshot()
})
