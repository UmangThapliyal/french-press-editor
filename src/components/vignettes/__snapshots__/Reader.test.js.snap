// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Reader without crashing, matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <_default />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "38",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          "",
        ],
      ],
      "node": null,
      "parent": Object {
        "getNextBlock": [Function],
      },
    },
    "ref": null,
    "rendered": Array [
      "",
    ],
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "38",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            "",
          ],
        ],
        "node": null,
        "parent": Object {
          "getNextBlock": [Function],
        },
      },
      "ref": null,
      "rendered": Array [
        "",
      ],
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
